New protocol documentation. (#380)
LiskHQ/lisk-docs: modules/ROOT/pages/6-appendix.adoc

== Appendix

=== [#index-serialization-1]#Serialization#

> The Lisk protocol specifies in [1] how to serialize objects used by the protocol. Generating a binary message from a JavaScript object is done following a stricter version of the https://developers.google.com/protocol-buffers/docs/encoding[Protocol Buffers] specification. Objects to be encoded have their format specified in a JSON schema. The schemas for blocks, transactions or accounts can be found in [<Block schema LIP>[2]], [<Transaction schema LIP>3] or [<Account schema LIP>4], respectively.

Reviewer: changed to LIP-00xx. We will have to update this once the LIPs have a number.

=== Signature Scheme

> The Lisk blockchain uses the Ed25519 signature scheme (see [https://en.wikipedia.org/wiki/Digital_signature[5]] and [https://ed25519.cr.yp.to/[6]]). The signing procedure uses the message and the private key of the signer to generate 64 bytes, called the signature. Other users can then verify this signature with respect to the message and the public key of the signer.

Reviewer: yes, changed.

On the Lisk blockchain the procedure for signing blocks and transactions is as follows:

. Serialize the object into the corresponding binary message.

> . Prepend the 32 bytes network identifier to the binary message (see [https://github.com/LiskHQ/lips/blob/master/proposals/lip-0009.md#specification[7]] and [https://github.com/LiskHQ/lips/blob/master/proposals/lip-0024.md#update-to-the-block-header-signing-procedure[8]]). The network identifier  uniquely binds a transaction to one network. Different blockchains created with the Lisk SDK should use different network identifiers to avoid transaction replay between chains.

Reviewer: added

. Compute the signature of the output of step 2 as described in the ed25519 specifications.

image::../assets/images/generating_signature.png[generating_signature]

==== Verifying Signatures
The network identifier is not transmitted with the binary message. The procedure to verify transactions is therefore:

> . Remove the signature from the object received.

Reviewer: This is on purpose. This section doesn't really talk about transmitting objects. So I consider you have an object, in some form. 
Maybe we could add a section about transmitting objects? I think it will probably wait for the technical documentation.

> . Remove the signature from the object received.

Reviewer 2: should be ok as it is

. Serialize the output of step 1 to a binary message.
. Prepend the network identifier to the binary message.
. Verify the output of step 3 with respect to the signature and the signer public key.

=== Object ID
The Lisk ecosystem uses IDs to identify objects, like blocks and transactions. The ID is obtained by serializing the object, including the signature, and taking the https://en.wikipedia.org/wiki/SHA-2[SHA-256] hash of this binary message.

image::../assets/images/generating_ID.png[generating_ID]

=== User Interface

> ==== Account Creation

Reviewer: changed to "Key Pair and Address Creation"

> To create a new account, first a 12-words passphrase from the BIP39 wordlist [1] is generated. The passphrase is then hashed using SHA-256 into a 256-bit string and the resulting hash is used as seed to generate the [#index-private_key-1]#*private key*# and the [#index-public_key-1]#*public key*# using Ed25519 signature scheme. Finally, the [#index-address-1]#address# is generated by taking the first 20 bytes of the public key.

Reviewer: added

> To create a new account, first a 12-words passphrase from the BIP39 wordlist [1] is generated. The passphrase is then hashed using SHA-256 into a 256-bit string and the resulting hash is used as seed to generate the [#index-private_key-1]#*private key*# and the [#index-public_key-1]#*public key*# using Ed25519 signature scheme. Finally, the [#index-address-1]#address# is generated by taking the first 20 bytes of the public key.

Reviewer: true the entropy is not high enought.
But I don't think this documentation is about becoming responsible. It is done like this and we should say so in the documentation.
We should definitely go up to 16 words, but I don't think this is on the agenda for now.

> To create a new account, first a 12-words passphrase from the BIP39 wordlist [1] is generated. The passphrase is then hashed using SHA-256 into a 256-bit string and the resulting hash is used as seed to generate the [#index-private_key-1]#*private key*# and the [#index-public_key-1]#*public key*# using Ed25519 signature scheme. Finally, the [#index-address-1]#address# is generated by taking the first 20 bytes of the public key.

Reviewer: changed.
Lisk elements is doing it as stated here. Except that the address is generated from the first 8 bytes reversed. We only document the future 20 bytes address here.

> To create a new account, first a 12-words passphrase from the BIP39 wordlist [1] is generated. The passphrase is then hashed using SHA-256 into a 256-bit string and the resulting hash is used as seed to generate the [#index-private_key-1]#*private key*# and the [#index-public_key-1]#*public key*# using Ed25519 signature scheme. Finally, the [#index-address-1]#address# is generated by taking the first 20 bytes of the public key.

Reviewer: changed

image::../assets/images/InfographicsV1/Infographic4.png[keyCreationDiagram]

==== User-Friendly Address
We display addresses in a user-friendly representation including a checksum and the “lsk” prefix. The procedure to obtain the [#index-user_friendly_address-1]#*user-friendly address*# from the address is as follows:

> . Calculate the checksum of the 20 byte address of the account using BCH code [https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki[1]]. This step gives a protection against accidental typing mistakes.

Reviewer: added

> . Calculate the checksum of the 20 byte address of the account using BCH code [https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki[1]]. This step gives a protection against accidental typing mistakes.

Reviewer 3: ```suggestion
. Calculate the checksum of the 20 bytes address of the account using BCH code [https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki[1]]. This step gives a protection against accidental typing mistakes.
```
not sure

Reviewer: will leave it

. Concatenate the the 20 byte address and the output of step 1.
. Encode the output of step 2 in a custom base32 format. We use lower-case letters and digits, but we remove the characters i, l, 1 and 0 for usability reasons.

> . Add the prefix lsk to the output of step 3. This means that all Lisk addresses used in the Lisk wallet start with lsk and that the final address is 41 characters long. 

Reviewer: added

The network identifier  uniquely binds a transaction or a block to one network. 
Different blockchains created with the Lisk SDK should use different network identifiers to avoid transaction replay between chains.
. Compute the signature of the output of step 2 above, as described in the Ed25519 specifications.

image::../assets/images/generating_signature.png[generating_signature]

=== Verifying Signatures
The network identifier is not transmitted with the binary message. 
The procedure to verify transactions is therefore as follows:

. Remove the signature from the object received.
. Serialize the output of step 1 to a binary message.
. Prepend the network identifier to the binary message.
. Verify the output of step 3 with respect to the signature and the signer public key.

== Object ID
The Lisk ecosystem uses IDs to identify objects, such as blocks and transactions. 
The ID is obtained by serializing the object, including the signature, and taking the https://en.wikipedia.org/wiki/SHA-2[SHA-256] hash of this binary message.

image::../assets/images/generating_ID.png[generating_ID]

== User Interface

=== Key Pair and Address Creation

> To create a new address, first a 12-words passphrase from the https://github.com/bitcoin/bips/blob/master/bip-0039/english.txt[english BIP39 wordlist] is generated. 

Reviewer 2: Generating the 12-word passphrase is done as described in bib39: https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#generating-the-mnemonic. That means, the words are not arbitrarily chosen, but derived from 128-bit number that was randomly chosen.
We could write:
"To create a new address, a 12-word passphrase is generated following the specifications of BIP39 (link to https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#generating-the-mnemonic)."

Reviewer: yes, done

Different blockchains created with the Lisk SDK should use different network identifiers to avoid transaction replay between chains.
. Compute the signature of the output of step 2 above, as described in the Ed25519 specifications.

image::../assets/images/unif_diagrams/signatureProcess.png[signature_process]

image::../assets/images/unif_diagrams/signatureProcess2.png[signature_process2]

=== Verifying Signatures
The network identifier is not transmitted with the binary message. 
The procedure to verify transactions is therefore as follows:

. Remove the signature from the object received.
. Serialize the output of step 1 to a binary message.
. Prepend the network identifier to the binary message.
. Verify the output of step 3 with respect to the signature and the signer public key.

== Object ID
The Lisk ecosystem uses IDs to identify objects, such as blocks and transactions. 
The ID is obtained by serializing the object, including the signature, and taking the https://en.wikipedia.org/wiki/SHA-2[SHA-256] hash of this binary message.

image::../assets/images/unif_diagrams/objectID.png[generating_ID]

image::../assets/images/unif_diagrams/objectID2.png[generating_ID]

> == User Interface

Author: I think "Address" would be a better title because this section explains both the binary address used in the protocol as well as the base32 representation.

Reviewer: changed

=== Key Pair and Address Creation
To create a new address, first a 12-word passphrase is generated following the specifications of https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki#generating-the-mnemonic[BIP39].
The passphrase is then hashed using SHA-256 into a 256-bit string. The resulting hash is used as the seed to generate the [#index-private_key-1]#*private key*# and the [#index-public_key-1]#*public key*# using Ed25519 signature scheme. 
Finally, the [#index-address-1]#address# is generated by taking the first 20 bytes of the SHA-256 hash of the public key.

image::../assets/images/unif_diagrams/addressCreation.png[addressCreation]

image::../assets/images/unif_diagrams/addressCreation2.png[addressCreation]

> === User-Friendly Address

Author: As decided in the last meeting, the user-friendly address is supposed to be called "Base32 Address". I would therefore use "Base32 Address" or "Base32 Representation" as title with a personal preference towards the second.

Reviewer: changed

> We display addresses in a user-friendly representation including a checksum and the “lsk” prefix. 

Author: ```suggestion
We display addresses in a Base32 representation including a checksum and the “lsk” prefix. 
```

Reviewer: changed

> The procedure to obtain the [#index-user_friendly_address-1]#*user-friendly address*# from the address is as follows:

Author: ```suggestion
The procedure to obtain the [#index-user_friendly_address-1]#*Base32 address*# from the address is as follows:
```

Reviewer: changed

. Calculate a 30-bit checksum of the 20-byte address of the account using a BCH code, see https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki[bip-0173]. This step provides protection against accidentally typing mistakes.
. Concatenate the 20 byte address and the output of step 1.

> . Encode the output of step 2 in a custom base32 format. Lower-case letters and digits are used, the characters i, l, 1 and 0 are removed for usability reasons.

Author: ```suggestion
. Encode the output of step 2 in a custom Base32 format. Lower-case letters and digits are used, the characters i, l, 1 and 0 are removed for usability reasons.
```

Author: I think Base32 should be capitalized as it is a proper name https://en.wikipedia.org/wiki/Base32.

Reviewer: done

> . Add the prefix "lsk" to the output of step 3. This ensures that all Lisk addresses used in the Lisk wallet start with "lsk" and that the final address is 41 characters long. 

Author: ```suggestion
. Add the prefix "lsk" to the output of step 3. This ensures that the Base32 addresses used in the Lisk wallet start with "lsk" and that the final address is 41 characters long. 
```

Reviewer: changed

> As an example, the address `0xc247a42e09e6aafd818821f75b2f5b0de47c8235` corresponds to the user-friendly address `lsk24cd35u4jdq8szo3pnsqe5dsxwrnazyqqqg5eu`.

Author: ```suggestion
As an example, the address `0xc247a42e09e6aafd818821f75b2f5b0de47c8235` has the Base32 representation `lsk24cd35u4jdq8szo3pnsqe5dsxwrnazyqqqg5eu`.
```

Reviewer: changed

= Appendix

== [#index-serialization-1]#Serialization#
The Lisk protocol specifies in [1] how to serialize objects used by the protocol. 
Generating a binary message from a JavaScript object is performed by following a stricter version of the https://developers.google.com/protocol-buffers/docs/encoding[Protocol Buffers] specification. 
Objects to be encoded have their format specified in a JSON schema. 
The schemas for blocks, transactions or accounts can be found in [<Block schema LIP>[LIP-00xx]], [<Transaction schema LIP>[LIP-00xx]] or [<Account schema LIP>[LIP-00xx]], respectively.

== Signature Scheme

> The Lisk protocol uses the Ed25519 signature scheme, please see the signatures page on https://en.wikipedia.org/wiki/Digital_signature[Wikipedia] and the specification page for https://ed25519.cr.yp.to/[Ed225519]). 

Reviewer 2: > please see the signatures page on https://en.wikipedia.org/wiki/Digital_signature[Wikipedia] and the specification page for https://ed25519.cr.yp.to/[Ed225519])

there is a closing bracket, but not an opening one.

Reviewer: removed, thanks

> The signing procedure uses the message and the private key of the signatory to generate 64 bytes, called the signature. 

Reviewer 2: > signatory

I think this term is not common at all in the context of digital signatures. But it's also not super important.

Reviewer: I agree. This was changed following one of Chris comments.
I will leave it for now, and see if more comments ask to change it.

The ID is obtained by serializing the object, including the signature, and taking the https://en.wikipedia.org/wiki/SHA-2[SHA-256] hash of this binary message.

image::../assets/images/generating_ID.png[generating_ID]

== User Interface

=== Key Pair and Address Creation
To create a new address, first a 12-words passphrase from the https://github.com/bitcoin/bips/blob/master/bip-0039/english.txt[english BIP39 wordlist] is generated. 
The passphrase is then hashed using SHA-256 into a 256-bit string. The resulting hash is used as the seed to generate the [#index-private_key-1]#*private key*# and the [#index-public_key-1]#*public key*# using Ed25519 signature scheme. 
Finally, the [#index-address-1]#address# is generated by taking the first 20 bytes of the SHA-256 hash of the public key.

image::../assets/images/InfographicsV1/Infographic4.png[keyCreationDiagram]

=== User-Friendly Address
We display addresses in a user-friendly representation including a checksum and the “lsk” prefix. 
The procedure to obtain the [#index-user_friendly_address-1]#*user-friendly address*# from the address is as follows:

> . Calculate a 30-bit checksum of the 20 byte address of the account using a BCH code, see https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki[bip-0173]. This step provvides protection against accidentally typing mistakes.

Reviewer 2: ```suggestion
. Calculate a 30-bit checksum of the 20 byte address of the account using a BCH code, see https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki[bip-0173]. This step provides protection against accidentally typing mistakes.
```

Reviewer: changed
